File: threading/python/02_reader_writer_lock.py
"""
QUESTION 2: Reader-Writer Lock (Rubrik/Glean Favorite)
======================================================

Problem Statement:
------------------
Implement a read-write lock that allows:
- Multiple readers to access shared resource simultaneously
- Only one writer at a time with exclusive access
- No readers while a writer is writing

Companies: Rubrik (storage systems), Glean (search indexing), Databricks (data processing)

Key Concepts:
- Shared vs Exclusive locks
- Reader preference vs Writer preference
- Starvation prevention
"""

import threading
import time
import random


class ReadWriteLock:
    """
    Read-Write Lock Implementation (Reader Preference)
    
    EXPLANATION:
    ------------
    This implementation gives preference to readers - as long as readers are active,
    new readers can acquire the lock even if writers are waiting.
    
    TRADE-OFF: This can cause writer starvation!
    
    State tracking:
    - readers: Count of active readers
    - writer: Boolean indicating if a writer is active
    - read_lock: Lock for updating reader count
    - write_lock: Lock for exclusive write access
    """
    
    def __init__(self):
        self.readers = 0
        self.read_lock = threading.Lock()  # Protects reader count
        self.write_lock = threading.Lock()  # Ensures exclusive write access
    
    def acquire_read(self):
        """
        Acquire read lock.
        
        CRITICAL INSIGHT:
        First reader blocks writers by acquiring write_lock.
        Subsequent readers just increment the count.
        """
        with self.read_lock:
            self.readers += 1
            if self.readers == 1:
                # First reader blocks writers
                self.write_lock.acquire()
    
    def release_read(self):
        """Release read lock."""
        with self.read_lock:
            self.readers -= 1
            if self.readers == 0:
                # Last reader allows writers
                self.write_lock.release()
    
    def acquire_write(self):
        """Acquire exclusive write lock."""
        self.write_lock.acquire()
    
    def release_write(self):
        """Release write lock."""
        self.write_lock.release()


class ReadWriteLockFair:
    """
    Fair Read-Write Lock (No Starvation)
    
    EXPLANATION:
    ------------
    Uses an additional lock to ensure fairness.
    When a writer is waiting, new readers must wait too.
    """
    
    def __init__(self):
        self.readers = 0
        self.read_lock = threading.Lock()
        self.write_lock = threading.Lock()
        self.order_lock = threading.Lock()  # Ensures FIFO ordering
    
    def acquire_read(self):
        with self.order_lock:  # Wait in line
            with self.read_lock:
                self.readers += 1
                if self.readers == 1:
                    self.write_lock.acquire()
    
    def release_read(self):
        with self.read_lock:
            self.readers -= 1
            if self.readers == 0:
                self.write_lock.release()
    
    def acquire_write(self):
        self.order_lock.acquire()  # Block new readers
        self.write_lock.acquire()
        self.order_lock.release()
    
    def release_write(self):
        self.write_lock.release()


# Shared resource
shared_data = {"value": 0}
rw_lock = ReadWriteLockFair()


def reader(reader_id: int, iterations: int):
    """Reader thread function"""
    for _ in range(iterations):
        rw_lock.acquire_read()
        try:
            print(f"Reader {reader_id} reading: {shared_data['value']}")
            time.sleep(random.uniform(0.05, 0.1))
        finally:
            rw_lock.release_read()
        time.sleep(random.uniform(0.01, 0.05))


def writer(writer_id: int, iterations: int):
    """Writer thread function"""
    for i in range(iterations):
        rw_lock.acquire_write()
        try:
            shared_data['value'] += 1
            print(f"Writer {writer_id} wrote: {shared_data['value']}")
            time.sleep(random.uniform(0.1, 0.2))
        finally:
            rw_lock.release_write()
        time.sleep(random.uniform(0.05, 0.1))


def main():
    threads = []
    # 3 readers, 2 writers
    for i in range(3):
        threads.append(threading.Thread(target=reader, args=(i, 5)))
    for i in range(2):
        threads.append(threading.Thread(target=writer, args=(i, 3)))
    
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    
    print(f"Final value: {shared_data['value']}")


if __name__ == "__main__":
    main()

File: threading/cpp/02_reader_writer_lock.cpp
/*
 * QUESTION 2: Reader-Writer Lock (Rubrik/Glean Favorite)
 * =======================================================
 * 
 * Problem: Implement a lock allowing multiple readers OR one exclusive writer.
 * 
 * Key Concepts: std::shared_mutex (C++17), custom RW lock implementation
 * 
 * Compile: g++ -std=c++17 -pthread 02_reader_writer_lock.cpp -o rw_lock
 */

#include <iostream>
#include <thread>
#include <mutex>
#include <shared_mutex>
#include <condition_variable>
#include <vector>
#include <chrono>

// Custom Reader-Writer Lock (Reader Preference)
class ReadWriteLock {
    /*
     * EXPLANATION:
     * - First reader acquires write_mutex (blocks writers)
     * - Additional readers increment count
     * - Last reader releases write_mutex
     * - WARNING: Can starve writers!
     */
private:
    std::mutex read_mutex_;
    std::mutex write_mutex_;
    int readers_ = 0;

public:
    void lock_read() {
        std::lock_guard<std::mutex> lock(read_mutex_);
        if (++readers_ == 1) {
            write_mutex_.lock();  // First reader blocks writers
        }
    }

    void unlock_read() {
        std::lock_guard<std::mutex> lock(read_mutex_);
        if (--readers_ == 0) {
            write_mutex_.unlock();  // Last reader allows writers
        }
    }

    void lock_write() {
        write_mutex_.lock();
    }

    void unlock_write() {
        write_mutex_.unlock();
    }
};

// Fair Reader-Writer Lock (No Starvation)
class FairReadWriteLock {
    /*
     * EXPLANATION:
     * order_mutex ensures FIFO ordering.
     * Writers block new readers when waiting.
     */
private:
    std::mutex read_mutex_;
    std::mutex write_mutex_;
    std::mutex order_mutex_;
    int readers_ = 0;

public:
    void lock_read() {
        std::lock_guard<std::mutex> order_lock(order_mutex_);
        std::lock_guard<std::mutex> lock(read_mutex_);
        if (++readers_ == 1) {
            write_mutex_.lock();
        }
    }

    void unlock_read() {
        std::lock_guard<std::mutex> lock(read_mutex_);
        if (--readers_ == 0) {
            write_mutex_.unlock();
        }
    }

    void lock_write() {
        order_mutex_.lock();
        write_mutex_.lock();
        order_mutex_.unlock();
    }

    void unlock_write() {
        write_mutex_.unlock();
    }
};

// Using C++17 std::shared_mutex (Recommended in production)
class SharedData {
private:
    mutable std::shared_mutex mutex_;
    int value_ = 0;

public:
    int read() const {
        std::shared_lock<std::shared_mutex> lock(mutex_);  // Multiple readers OK
        return value_;
    }

    void write(int value) {
        std::unique_lock<std::shared_mutex> lock(mutex_);  // Exclusive access
        value_ = value;
    }
};

SharedData shared_data;
FairReadWriteLock rw_lock;

void reader(int id) {
    for (int i = 0; i < 5; ++i) {
        rw_lock.lock_read();
        std::cout << "Reader " << id << " read: " << shared_data.read() << std::endl;
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
        rw_lock.unlock_read();
    }
}

void writer(int id) {
    for (int i = 0; i < 3; ++i) {
        rw_lock.lock_write();
        shared_data.write(shared_data.read() + 1);
        std::cout << "Writer " << id << " wrote: " << shared_data.read() << std::endl;
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        rw_lock.unlock_write();
    }
}

int main() {
    std::vector<std::thread> threads;
    
    for (int i = 0; i < 3; ++i) threads.emplace_back(reader, i);
    for (int i = 0; i < 2; ++i) threads.emplace_back(writer, i);
    
    for (auto& t : threads) t.join();
    
    std::cout << "Final value: " << shared_data.read() << std::endl;
    return 0;
}

